src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -282,166 +292,195 @@ struct TreeVisitor<'tree> {
 /// Whether to continue exploring the children recursively or not.
 enum ContinueTraversal {
     Recurse,
-    SkipChildren,
+    SkipSelfAndChildren,
 }
 
 /// Stack of nodes left to explore in a tree traversal.
-struct TreeVisitorStack<NodeApp, ErrHandler> {
+struct TreeVisitorStack<NodeContinue, NodeApp, ErrHandler> {
     /// Identifier of the original access.
     initial: UniIndex,
+    /// Function describing whether to continue at a tag.
+    /// This is only invoked for foreign accesses.
+    f_continue: NodeContinue,
     /// Function to apply to each tag.
     f_propagate: NodeApp,
     /// Handler to add the required context to diagnostics.
     err_builder: ErrHandler,
     /// Mutable state of the visit: the tags left to handle.
     /// Every tag pushed should eventually be handled,
     /// and the precise order is relevant for diagnostics.
-    stack: Vec<(UniIndex, AccessRelatedness)>,
+    /// Since the traversal is bottom-up, we need to remember
+    /// whether we're here initially (false) or after visiting all
+    /// children (true). The bool indicates this.
```

Comment:
If it is bottom-up, the access would always be after visiting all children, right?

Reply:
Well, yes, but actually no. There is infrastructure for having the visitation skip entire subtrees if it would not change anything. This is done by remembering the last visit, and if the last visit was a foreign read, and the next visit is a foreign read, then it skips that.

(Note: I just realized that this may be unsound to do in combination with `lazy`, let me check.. But that's the motivation for visiting everything twice).

Reply:
Indeed, skipping subtree updates the way we do currently is, eh, wrong and broken :grimacing:  See https://github.com/rust-lang/miri/issues/3846. We should find a solution to that before we continue here.

Reply:
And independently of that infrastructure, you must first _get to_ the child to visit. For this you need to walk down.

Reply:
This comment makes a lot more sense after reading the other comments further down that explain the actual iteration order. :)  I wouldn't quite call this "bottom-up".